src/pages/library/datasetBuilder/DatasetBuilder.ts

```diff
@@ -0,0 +1,419 @@
+                                        selectedDatasetBuilderObjectSets
+                                      )
+                                );
+                              },
+                            },
+                            [label({ style: { paddingLeft: '0.5rem' } }, [value.name])]
+                          ),
+                        ]
+                      );
+                    }, Utils.toIndexPairs(datasetBuilderObjectSet.values)),
+                  ]),
+                Utils.toIndexPairs(
+                  _.filter(
+                    (datasetBuilderObjectSet) => datasetBuilderObjectSet.values?.length > 0,
+                    datasetBuilderObjectSets
+                  )
+                )
+              )
+            )
+          : div([placeholder]),
+      ]
+    ),
+  ]);
+};
+
+export const CreateCohortModal = ({
+  onDismiss,
+  onStateChange,
+}: {
+  onDismiss: () => void;
+  onStateChange: OnStateChangeType;
+}) => {
+  const [cohortNameTouched, setCohortNameTouched] = useState(false);
+  const [cohortName, setCohortName] = useState('');
+
+  const errors =
+    cohortNameTouched &&
+    validate(
+      { cohortName },
+      {
+        cohortName: {
+          presence: {
+            allowEmpty: false,
+          },
+          exclusion: {
+            within: _.map((cohort: Cohort) => cohort.name, datasetBuilderCohorts.get()),
+            message: 'already exists',
+          },
+        },
+      }
+    );
+
+  const createCohort = (cohortName) => {
+    // Once state is typed, the ts-ignore should go away
+    // @ts-ignore
+    datasetBuilderCohorts.set(datasetBuilderCohorts.get().concat({ name: cohortName }));
+    onStateChange('cohort-editor');
+  };
+
+  return h(
+    Modal,
+    {
+      onDismiss,
+      title: 'Create a new cohort',
+      okButton: h(
+        ButtonPrimary,
+        {
+          onClick: () => {
+            createCohort(cohortName);
+            onDismiss();
+          },
+          disabled: !cohortNameTouched || errors?.cohortName,
+        },
+        ['Create cohort']
+      ),
+    },
+    [
+      h(StringInput, {
+        title: 'Cohort name',
+        onChange: (value) => {
+          !cohortNameTouched && setCohortNameTouched(true);
+          setCohortName(value);
+        },
+        value: cohortName,
+        errors: errors?.cohortName,
+        placeholder: 'Enter the cohort name',
+        required: true,
+      }),
+    ]
+  );
+};
+
+export const CohortSelector = ({
+  selectedCohorts,
+  onChange,
+  onStateChange,
+}: {
+  selectedCohorts: DatasetBuilderObjectSet<Cohort>[];
+  onChange: (cohorts: DatasetBuilderObjectSet<Cohort>[]) => void;
+  onStateChange: OnStateChangeType;
+}) => {
+  const [creatingCohort, setCreatingCohort] = useState(false);
+
+  return h(Fragment, [
+    h(DatasetBuilderSelector as React.FC<DatasetBuilderSelectorProps<Cohort>>, {
```

Comment:
Is there a way to add a return type to `DatasetBuilderSelector` to avoid the need for `as` here?